hpx/lcos/dist_object.hpp

```diff
@@ -0,0 +1,566 @@
+        hpx::id_type get_basename_helper(int idx)
+        {
+            if (!basename_list[idx])
+            {
+                basename_list[idx] =
+                    hpx::find_from_basename(base_ + std::to_string(idx), idx)
+                        .get();
+            }
+            return basename_list[idx];
+        }
+        void basename_registration_helper(std::string base)
+        {
+            base_unpacked = base + std::to_string(hpx::get_locality_id());
+            hpx::register_with_basename(
+                base + std::to_string(hpx::get_locality_id()), this->get_id());
+            basename_list.resize(hpx::find_all_localities().size());
+        }
+        /// \endcond
+    };
+
+    /// The dist_object is a single logical object partitioned over a set of
+    /// localities/nodes/machines, where every locality shares the same global
+    /// name locality for the distributed object (i.e. a universal name), but
+    /// owns its local value. In other words, local data of the distributed
+    /// object can be different, but they share access to one another's data
+    /// globally.
+    template <typename T, construction_type C>
+    class dist_object<T&, C>
+      : hpx::components::client_base<dist_object<T&>,
+            server::dist_object_part<T&>>
+    {
+        /// \cond NOINTERNAL
+        typedef hpx::components::client_base<dist_object<T&>,
+            server::dist_object_part<T&>>
+            base_type;
+
+        typedef typename server::dist_object_part<T&>::data_type data_type;
+
+    private:
+        template <typename Arg>
+        static hpx::future<hpx::id_type> create_server(Arg& value)
+        {
+            return hpx::local_new<server::dist_object_part<T&>>(value);
+        }
+        /// \endcond
+    public:
+        /// Creates a dist_object in every locality
+        ///
+        /// A dist_object \a base_name is created through default constructor.
+        dist_object() {}
+
+        /// Creates a dist_object in every locality with a given base_name string,
+        /// data, and a construction_type. This constructor of the dist_object
+        /// wraps an existing local instance and thus is internally referring to
+        /// the local instance.
+        ///
+        /// \param construction_type The construction_type in the template parameters
+        /// accepts either Meta_Object, and it is set to All_to_All by defalut
+        /// The Meta_Object option provides meta object registration in the root
+        /// locality and meta object is essentailly a table that can find the
+        /// instances of dist_object in all localities. The All_to_All option only
+        /// locally holds the client and server of the dist_object.
+        /// \param base_name The name of the dist_object, which should be a unique
+        /// string across the localities
+        /// \param data The data of the type T of the dist_object
+        dist_object(std::string base, data_type data)
```

Comment:
There should also be a constructor that takes a vector of localities so that an object can be created on a subset of all

Reply:
Now it can accept a list of sub-localities.
```cpp
        distributed_object(std::string base, data_type const& data, std::vector<size_t> sub_localities = all_localities())
```
https://github.com/STEllAR-GROUP/hpx/blob/b1c8f32ae0627852cf02f96013a70b443f03a066/hpx/lcos/distributed_object.hpp#L301